crates/controller/src/hud/details.rs

```diff
@@ -1,37 +1,113 @@
+            .add_system(
+                update
+                    .in_base_set(GameSet::PostUpdate)
+                    .run_if(in_state(GameState::Playing)),
+            );
     }
 }
 
-fn spawn_details(mut commands: Commands) {
-    commands.spawn((
-        NodeBundle {
-            style: Style {
+#[derive(Resource)]
+struct DetailsText(Entity);
+
+fn setup(mut commands: GuiCommands) {
+    let node = commands
+        .spawn((
+            NodeBundle {
+                style: Style {
+                    size: Size {
+                        width: Val::Percent(20.),
+                        height: Val::Percent(30.),
+                    },
+                    position_type: PositionType::Absolute,
+                    position: UiRect::new(
+                        Val::Percent(0.),
+                        Val::Percent(20.),
+                        Val::Percent(70.),
+                        Val::Percent(100.),
+                    ),
+                    ..default()
+                },
+                background_color: HUD_COLOR.into(),
+                ..default()
+            },
+            DespawnOnGameExit,
+            InteractionBlocker,
+        ))
+        .id();
+    let details_text = commands
+        .spawn_body_text(
+            OuterStyle {
                 size: Size {
                     width: Val::Percent(20.),
                     height: Val::Percent(30.),
                 },
-                position_type: PositionType::Absolute,
-                position: UiRect::new(
-                    Val::Percent(0.),
-                    Val::Percent(20.),
-                    Val::Percent(70.),
-                    Val::Percent(100.),
-                ),
-                ..default()
+                margin: UiRect {
+                    left: Val::Px(10.),
+                    right: Val::Px(10.),
+                    top: Val::Px(10.),
+                    bottom: Val::Px(10.),
+                },
             },
-            background_color: HUD_COLOR.into(),
-            ..default()
-        },
-        DespawnOnGameExit,
-        InteractionBlocker,
-    ));
+            "",
+        )
+        .id();
+    commands.entity(node).add_child(details_text);
+
+    commands.insert_resource(DetailsText(details_text));
+}
+
+fn update(
+    ui: Res<DetailsText>,
+    selected: Query<Entity, With<Selected>>,
+    battery: Query<&Battery>,
+    mut text_ops: BodyTextOps,
+) {
+    if selected.is_empty() {
+        text_ops
+            .set_text(ui.0, "".into())
+            .expect("Failed to set text of details");
+        return;
+    }
+
+    let mut battery_total = 0;
+    let mut battery_max = 0;
+
+    for entity in selected.iter() {
+        if let Ok(battery) = battery.get(entity) {
+            battery_total += battery.energy().floor() as i64;
+            battery_max += battery.capacity().floor() as i64;
+        }
+    }
+
+    if battery_max == 0 {
+        return;
+    }
+
+    let text = format!(
```

Comment:
With:

```rust
const PREFIXES: [&'static str; 5] = ["T", "G", "M", "k", ""];
```

and

```rust
fn format_units(value: f64, units: &str) -> String {
    for i in (0..PREFIXES.len()).rev() {
        let coeff = 1000f64.powi(i as i32);
        if value > coeff {
            let rounded = most_significant(value / coeff);
            return format!("{}{}{units}", rounded, PREFIXES[i]);
        }
        
    }
    
    unreachable!("TODO");
}

fn most_significant(value: f64) -> f64 {
    if value == 0.0 {
        0.0
    } else {
        let d = 3 - value.abs().log10().ceil() as i32;
        let p = 10f64.powi(d);
        (value * p).round() / p
    }
}
```

we can use pretty `format_units(battery_max, "J")` and `format_units(battery_total, "J")`.

Reply:
Very pretty!